File: advent-of-code/2025/day-2-gift-shop/normal/n-d02-12.kt
fun isDouble(n: Long): Boolean {
    val s = n.toString()
    if (s.length % 2 != 0) return false
    return s.substring(0, s.length / 2) == s.substring(s.length / 2)
}

fun isRepeat(n: Long): Boolean {
    val s = n.toString()
    for (k in 1 until s.length) {
        if (s.length % k == 0) {
            val pat = s.substring(0, k)
            if (pat. repeat(s.length / k) == s) return true
        }
    }
    return false
}

fun main() {
    val data = generateSequence(::readLine).joinToString("").trim()
    var p1 = 0L
    var p2 = 0L
    for (r in data.split(",")) {
        if (r.isBlank()) continue
        val (lo, hi) = r.split("-").map { it.toLong() }
        for (n in lo..hi) {
            if (isDouble(n)) p1 += n
            if (isRepeat(n)) p2 += n
        }
    }
    println(p1)
    println(p2)
}

File: advent-of-code/2025/day-2-gift-shop/normal/n-d02-12.rs
use std::io::{self, Read};

fn is_double(n: i64) -> bool {
    let s = n.to_string();
    if s.len() % 2 != 0 { return false; }
    let h = s.len() / 2;
    &s[..h] == &s[h..]
}

fn is_repeat(n: i64) -> bool {
    let s = n.to_string();
    for k in 1..s.len() {
        if s.len() % k == 0 {
            let pat = &s[..k];
            if pat.repeat(s.len() / k) == s { return true; }
        }
    }
    false
}

fn main() {
    let mut input = String::new();
    io::stdin().read_to_string(&mut input).unwrap();
    let data: String = input.chars().filter(|&c| c != '\n').collect();
    let (mut p1, mut p2) = (0i64, 0i64);
    for r in data.split(',') {
        if r.trim().is_empty() { continue; }
        let parts: Vec<i64> = r.split('-').map(|x| x.parse().unwrap()).collect();
        for n in parts[0]..=parts[1] {
            if is_double(n) { p1 += n; }
            if is_repeat(n) { p2 += n; }
        }
    }
    println!("{}\n{}", p1, p2);
}
